### DISCUSSION TITLE
Multi-tenancy of framework

### DISCUSSION THREAD

**Discussion by arnefromsa** [aws-solutions-library-samples/guidance-for-aws-sustainability-insights-framework#17] [Q&A] [answered]
Goodday, 
We have been looking at the multi-tenancy of the SIF framework, but cannot find documentation or a diagram that shows what components are duplicated, and what will be centrally managed in the framework.  

Could you possibly point us in the right direction as I could not find exactly the information from an architecture perspective.

The only reference I could find in the [design documentation is found here](https://github.com/aws-solutions-library-samples/guidance-for-aws-sustainability-insights-framework/blob/main/docs/design.md).  

> SIF is designed to support multi-tenant deployments. Separate tenants of SIF can be deployed per AWS region and per environment (development, test, prod). In a multi-tenant deployment a separate set of infrastructure is deployed for each tenant in each environment.

**Comment by williamsia** [comment, marked as answer]
Hi Arne, you're right I don't think we have any section that explicitly describes how multi tenancy work in SIF, I will raise a ticket with the team to add this documentation for the future release. But here is the overview of how the multi-tenancy works in SIF. 

Basically the modules that are supported for multi tenancy are `referenceDatasets`, `calculations` and `impacts`. The operation that are supported are `GET` only, so you could not modify resources on another tenant. 

Let use 2 tenant for example, `Tenant A` ( the calling tenant )  and `Tenant B` ( the shared tenant )  

When you deploy Tenant B you specified which groups (as a cdk context argument) in this tenant that available for other tenant to use 

Below is the simplified command line to allow the group `/shared` to be called by external tenant

```shell
cdk deploy --app cdk.out -c tenantId=tenantB  -c externallySharedGroupIds=/shared --all
```

When you deploy Tenant A (the calling tenant), you specify which shared tenant ( and the group on that tenant) that TenantA allow to invoke 

The simplified command below allow tenant A to retrieve resources (referenceDatasets, impacts, calculations) in tenant B on group `/shared`

```shell
cdk deploy -c tenantId=tenantA -c outgoingTenantPaths=tenantB:/shared --all
```

To retrieve resources in TenantB from TenantA, what you have to do is to specify 2 http headers when you invoking the `TenantA` api endpoint.
1. `x-tenant` set this to `tenantB`
2. `x-groupcontextid` set this to `/shared`

When you invoke api endpoint in `TenantA` with this headers, it will proxy your call to retrieve resources in `TenantB`, hopefully this gives you some high level idea of how it works. 

In term of resources that get deployed, it will be the same in `TenantA` and `TenantB`, but both will share the Environment/Platform resources (e.g. ECS Cluster, RDS Cluster )
